### PR TITLE
storage/glusterfs: enable glusterfs deployment

### PR DESCRIPTION
**Pull request by raghavendra-talur**
We add a new target in Makefile "cluster-storage-deploy" which can be
called after "cluster-sync" to create a gluster backed storage backend
for kubevirt.

Changes:
a. Add 3 disks of size 500GiB on all nodes that are created.
b. install gluster client packages on host nodes
c. created a new dir for storage related stuff

TODO:
a. make the same changes for vagrant-openshift
b. make the topology file creation dynamic

Signed-off-by: Raghavendra Talur <rtalur@redhat.com>

### CODE REVIEW

**Comment by ovirt-infra** [comment]
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically.
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

**Comment by jarrpa** [comment]
Looking good thus far.  `gluster-deploy.sh` references some variable strings in files being run through `sed` that are not present in those ifles yet, but I imagine that's known. :)

**Comment by raghavendra-talur** [comment]
@jarrpa my mistake. Fixed and some more changes.

**Comment by jarrpa** [comment]
Looks good on my end. Should be sufficient for current purposes.

File: cluster/vagrant-kubernetes/setup_common.sh [hunk 1]
```diff
@@ -93,3 +94,7 @@ systemctl enable kubelet && systemctl start kubelet
 modprobe bridge
 sysctl -w net.bridge.bridge-nf-call-iptables=1
 sysctl -w net.bridge.bridge-nf-call-ip6tables=1
+
+# Storage
+yum install centos-release-gluster
+yum install --nogpgcheck -y glusterfs-fuse
```
Comment (nixpanic) [inline cluster/vagrant-kubernetes/setup_common.sh]: You should not need the `--nogpgcheck`. Packages are signed with the GPG-KEY that comes with the centos-release-gluster package.

File: cluster/vagrant-openshift/setup_common.sh [hunk 2]
```diff
@@ -32,3 +32,7 @@ yum install -y centos-release-openshift-origin
 yum install -y yum-utils ansible wget git net-tools bind-utils iptables-services bridge-utils bash-completion kexec-tools sos psacct docker
 systemctl start docker
 systemctl enable docker
+
+# Storage
+yum install centos-release-gluster
+yum install --nogpgcheck -y glusterfs-fuse
```
Comment (nixpanic) [inline cluster/vagrant-openshift/setup_common.sh]: Same for `--nogpgcheck` here.

File: cluster/vagrant-kubernetes/setup_common.sh (hunk 1 above)
Comment (rmohr) [reply cluster/vagrant-kubernetes/setup_common.sh]: Why do we need glusterfs-fuse on the machine? Can't we bring that with the container?

File: Vagrantfile [hunk 3]
```diff
@@ -55,6 +55,17 @@ Vagrant.configure(2) do |config|
       domain.nested = true  # enable nested virtualization
       domain.cpu_mode = "host-model"
 
+      # We need 3 bricks, let us create 500G disks for that.
+      # NOTE: if size is changed, edit storage/gluster-deploy.sh
+      "bcd".split("").each do |i|
```
Comment (rmohr) [inline Vagrantfile]: Could you explain why, we need dedicated disks and can't have something completely inside the vm?

File: storage/deploy.sh [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+#
+# Copyright 2017 Red Hat, Inc.
+#
+
+
+echo "Deploying Storage..."
+vagrant rsync master # if you do not use NFS
+
+# Gluster
+vagrant ssh master -c "sudo /sbin/iptables -A INPUT -p tcp -m state --state NEW -m tcp --dport 24007 -j ACCEPT"
```
Comment (rmohr) [inline storage/deploy.sh]: The firewall is disabled, why is this needed?

File: storage/deploy.sh [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+vagrant ssh master -c "sudo /sbin/iptables -A INPUT -p tcp -m state --state NEW -m tcp --dport 24008 -j ACCEPT"
+vagrant ssh master -c "sudo /sbin/iptables -A INPUT -p tcp -m state --state NEW -m tcp --dport 2222 -j ACCEPT"
+vagrant ssh master -c "sudo /sbin/iptables -A INPUT -p tcp -m state --state NEW -m multiport --dports 49152:49664 -j ACCEPT"
+vagrant ssh master -c "sudo /sbin/iptables -A INPUT -p tcp -m state --state NEW -m tcp --dport 24010 -j ACCEPT"
+vagrant ssh master -c "sudo /sbin/iptables -A INPUT -p tcp -m state --state NEW -m tcp --dport 3260 -j ACCEPT"
+vagrant ssh master -c "sudo /sbin/iptables -A INPUT -p tcp -m state --state NEW -m tcp --dport 111 -j ACCEPT"
+vagrant ssh master -c "cd /vagrant && sudo storage/gluster-deploy.sh"
```
Comment (rmohr) [inline storage/deploy.sh]: Can't we completely deploy gluster from within a container?

File: cluster/vagrant-kubernetes/setup_common.sh (hunk 1 above)
Comment (rmohr) [reply cluster/vagrant-kubernetes/setup_common.sh]: Is that used by the kubelet or by a flexvolume?

Comment (jarrpa) [reply cluster/vagrant-kubernetes/setup_common.sh]: kubelet. Kubernetes has a native GlusterFS mount plugin that uses glusterfs-fuse.

File: Vagrantfile (hunk 3 above)
Comment (rmohr) [reply Vagrantfile]: Can we have a container, which is part of the manifest, which creates three qcow2 disks on deployment and mounts them as nbd devices instead?

File: cluster/vagrant-kubernetes/setup_common.sh (hunk 1 above)
Comment (rmohr) [reply cluster/vagrant-kubernetes/setup_common.sh]: ok, then it makes sense to install  it. :+1:

File: storage/gluster-deploy.sh [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Copyright 2017 Red Hat, Inc.
+#
+
+
+PROVIDER=${PROVIDER:-vagrant-kubernetes}
```
Comment (rmohr) [inline storage/gluster-deploy.sh]: All that looks like it can be done from an init container, or directly by the cluster container in the manifest.

File: storage/deploy.sh (hunk 5 above)
Comment (rmohr) [reply storage/deploy.sh]: Could you clarify a little bit more why we need all these iptables rules are needed? AFAIK we have disabled the firewall.

File: storage/gluster-deploy.sh (hunk 6 above)
Comment (rmohr) [reply storage/gluster-deploy.sh]: ok now I get it. gk-deploy does some deployment in k8s. That all looks way too complicated.

File: Vagrantfile (hunk 3 above)
Comment (jarrpa) [reply Vagrantfile]: I don't understand where that would run, nor how that's any simpler/easier/cleaner than doing it in the Vagrantfile.

File: storage/deploy.sh (hunk 5 above)
Comment (jarrpa) [reply storage/deploy.sh]: I didn't write it, so unfortunately I can't explain why within this context.

**Comment by jarrpa** [comment]
@rmohr The gk-deploy script is all we have available for Kubernetes, and has been used successfully by many users even in production environments. We have plans to try and streamline the process further, but we're waiting for some other things to fall into place first.

File: Vagrantfile (hunk 3 above)
Comment (rmohr) [reply Vagrantfile]: The more we do inside the manifests, the easier it is to deploy it on differnt clusters. See for instance: https://github.com/rmohr/ceph-cinder-demo/blob/master/storage/templates/provisioner.yaml. This provisions ceph and cinder in a  single-node setup, without the need to touch the node in any way. For as long as a user can run provileged pods (for host network), this can be used in any k8s/openshift cluster.

That has a few advantages:
 * One does not need to rely on the "kubevirt" repo, just to develop with storage
 * One does not have create/change anything on the node to add storage, one just adds or removes manifests (no vagrant destroy, ...).
 * Of course that all is not for production, but we only need it to test code on top of gluster/cinder/ceph
...

I hope to see the same for gluster, so that we can simply add/remove gluster on arbitrary running clusters if needed.

Comment (jarrpa) [reply Vagrantfile]: How do you creat qcoe2 images and attach them to the VM from inside the VM?

Comment (rmohr) [reply Vagrantfile]: Out of my head I know of two ways on how to create and attach the files from within a container:
 * loopback
 * nbd

Loopback works in principle like this inside a container:

```bash
truncate disk.img --size 2G
mknod /dev/loop9 -m0660 b 7 9
losetup /dev/loop9 disk.img
losetup -a
```

For nbd make sure that you mount "/lib/modules" to the container, then do something like this:

```bash
modprobe nbd max_part=8
mknod /dev/nbd0 b 43 0
qemu-nbd --connect=/dev/nbd0 file.qcow2
```

Comment (jarrpa) [reply Vagrantfile]: Hmm... so this is done within the container and it changes the state of the host it's running on? And what are the differences, advantages, and/or disadvantages between the two methods?

Comment (rmohr) [reply Vagrantfile]: Maybe the loopback is more predictable and easier, since we don't need to load kernel modules.

It is a little bit hard for me to see what gk-deploy mechanism does with the topology, but I guess a gluster component is started on the node, which exposes these devices via a service? With that approach we can mount the devices inside the container (that this is also visible on the host is just a side-effect).

Is there then anything missing, to just provide a pre-created yaml manifest for development/testing without gk-deploy? Is there still something running on the node, or needs to be run on the node?

Comment (jarrpa) [reply Vagrantfile]: gk-deploy does a fair number of things that can't be encapsulated in a single YAML manifest. Many of the resources deployed have to be done in a specific order and not all at once. You also can't pre-populate the heketi database, so the topology load, volume creation, and database copy have to be performed every time.

I would recommend setting up a meeting to get a run-down of what gk-deploy does, how the different components operate, and what requirements are mandatory vs. configuration-specific.

Comment (rmohr) [reply Vagrantfile]: > I would recommend setting up a meeting to get a run-down of what gk-deploy does, how the different components operate, and what requirements are mandatory vs. configuration-specific.

We can talk about it, sure.

> gk-deploy does a fair number of things that can't be encapsulated in a single YAML manifest. Many of the resources deployed have to be done in a specific order and not all at once.

That's the same for ceph and cinder. There are two very simple ways to make sure that the order is right:
1. Let containers exit with a non-zero exit code if it can't connect to a prerequisite service
2. Let containers wait for "ready" files to appear in a shared directory (emptyDir).

Comment (jarrpa) [reply Vagrantfile]: Hmm... I'm not very well versed in container-thinking yet, it seems. I'd like to learn more about this. Probably in the same meeting as above? :)

File: storage/gluster-deploy.sh (hunk 6 above)
Comment (davidvossel) [reply storage/gluster-deploy.sh]: pulling in a 1000 line bash script is concerning.

Comment (jarrpa) [reply storage/gluster-deploy.sh]: It's what we have now. It's been verified and used in enterprise production environments. We're looking ahead to changing this, but it's not something we'll get to soon.

**Comment by rmohr** [comment]
Closing this for now. Feel free to open it again if you want to continue working on it.
